### ISSUE TITLE
Allow Final attributes declared in dataclass body without a default value

### ISSUE THREAD

**Issue by sisp** [python/mypy#10688]
**Feature**

It should be allowed to declare a read-only (final) dataclass attribute:

```python
@dataclass
class A:
    x: Final[str]  # error: Final name must be initialized with a value  [misc]
```

**Pitch**

I think it's a common scenario where a dataclass attribute should be read-only once the dataclass has been instantiated. It works fine with a default value

```python
@dataclass
class B:
    x: Final[str] = 'abc'  # ok
```

and with a regular class:

```python
class B:
    x: Final[str]  # ok
    
    def __init__(self, x: str) -> None:
        self.x = x
```

See the full example on the [mypy playground](https://mypy-play.net/?mypy=latest&python=3.9&gist=0aa3cc80982a596870798b658b8662b9).

**Comment by erictraut** [comment]
As a simple workaround, you could assign it a field descriptor.

```python
@dataclass
class A:
    x: Final[str] = field()
```

**Comment by sisp** [comment]
@erictraut Thanks for suggesting the workaround! A proper fix would still be nice.

**Comment by AlexWaygood** [comment]
Closing as a duplicate of #5608
